File: 2024/go/day01/main.go
package main

import (
	"aoc-shared/pkg/sharedcode"
	"aoc-shared/pkg/sharedstruct"
	"math"
	"os"
	"path/filepath"
	"runtime"
	"sort"
	"strconv"
	"strings"
)

func getCurrentDirectory() string {
	_, filename, _, _ := runtime.Caller(0)
	dirname := filepath.Dir(filename)
	return dirname
}

// Default Input path is current directory + example-input
var inputPath = filepath.Join(getCurrentDirectory(), "example-input")
var isUsingExample = true

func main() {
	// If another cmd argument has been passed, use that as the input path:
	if len(os.Args) > 1 {
		inputPath = os.Args[1]
		isUsingExample = false
	}

	var _, contents = sharedcode.ParseFile(inputPath)

	partOne(contents)
	partTwo(contents)
}

func partOne(contents []string) {
	combos := parseInput(contents)

	sort.Slice(combos[0], func(i, j int) bool {
		return combos[0][i] < combos[0][j]
	})

	sort.Slice(combos[1], func(i, j int) bool {
		return combos[1][i] < combos[1][j]
	})

	difference := int64(0)

	for i := 0; i < len(combos[0]); i++ {
		difference += int64(math.Abs(float64(combos[0][i] - combos[1][i])))
	}

	sharedstruct.PrintOutput(sharedstruct.Output{
		Day:   1,
		Part:  1,
		Value: difference,
	})
}

func partTwo(contents []string) {
	combos := parseInput(contents)

	sort.Slice(combos[0], func(i, j int) bool {
		return combos[0][i] < combos[0][j]
	})

	sort.Slice(combos[1], func(i, j int) bool {
		return combos[1][i] < combos[1][j]
	})

	similarityScore := int64(0)

	for i := 0; i < len(combos[0]); i++ {
		// We go through all of the lhs list, and multiply the number by the qnty in the rhs list
		qty := int64(0)
		currVal := combos[0][i]

		for j := 0; j < len(combos[1]); j++ {
			if combos[1][j] > currVal {
				break
			}
			if combos[1][j] == currVal {
				qty++
			}
		}

		similarityScore += qty * currVal
	}
	sharedstruct.PrintOutput(sharedstruct.Output{
		Day:   1,
		Part:  2,
		Value: similarityScore,
	})
}

func parseInput(contents []string) [2][]int64 {
	leftCombos := make([]int64, 0)
	rightCombos := make([]int64, 0)

	for _, line := range contents {
		var numOne, numTwo int64

		tmp := strings.Split(line, "   ")

		numOne, _ = strconv.ParseInt(tmp[0], 10, 64)
		numTwo, _ = strconv.ParseInt(tmp[1], 10, 64)

		leftCombos = append(leftCombos, numOne)
		rightCombos = append(rightCombos, numTwo)
	}

	return [2][]int64{leftCombos, rightCombos}
}

File: 2024/ts/day01/main.ts
import { fromFileUrl, dirname } from "jsr:@std/path";

// --- SETUP ---
const text = await readFile();
const [leftList, rightList] = parseInput(text);

leftList.sort((a, b) => a - b);
rightList.sort((a, b) => a - b);

// --- PART 1 ---
let difference = 0;

for (let i = 0; i < leftList.length; i++) {
  difference += Math.abs(leftList[i] - rightList[i]);
}

console.log({ difference });

// --- PART 2 ---
let similarityScore = 0;

leftList.forEach((val) => {
  let qty = 0;

  for (let j = 0; j < rightList.length; j++) {
    if (val === rightList[j]) {
      qty++;
    }
    if (rightList[j] > val) {
      break;
    }
  }

  similarityScore += qty * val;
});

console.log({ similarityScore });

async function readFile(): string {
  let path = dirname(fromFileUrl(import.meta.url)) + "/example-input";
  if (Deno.args.length > 0) {
    // If we pass an argument, use this as the input
    path = Deno.args[0];
  }

  return await Deno.readTextFile(path);
}

function parseInput(input: string): [number[], number[]] {
  let leftList = [] as number[];
  let rightList = [] as number[];
  input.split("\n").forEach((line) => {
    let [left, right] = line.split("   ").map(Number);

    leftList.push(left);
    rightList.push(right);
  });

  return [leftList, rightList];
}
